### PR TITLE
Add taunt command so players can taunt each other.

### PR DESCRIPTION
**Pull request by marin-hyatt**
When a player enters "@pongbot taunt @playername", the bot will announce that
@playername sucks at ping pong. Example:

@pongbot taunt @player2

@player1 said @player2 sucks at ping pong!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -197,6 +197,16 @@ gamebot cancel
...
 ```
 
+####  gamebot taunt &lt;opponent&gt; ... [with &lt;teammate&gt; ...] 
+
+Taunt another player.
+
+```
+gamebot taunt @WangHoe
+
+Victor Barna said Wang Hoe sucks at ping pong!
```
Comment (dblock) [inline README.md]: I think it should be "says", because this is happening right now. Victor Barna just sucks at ping pong. It's not going to change.

File: slack-gamebot/commands/taunt.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+        if taunter.registered?
+          arguments = match['expression'].split.reject(&:blank?) if match['expression']
+          arguments ||= []
+          taunt = ::Taunt.create_from_teammates_and_opponents!(client.owner, data.channel, taunter, arguments)
+          client.say(channel: data.channel, text: "#{taunt.taunters.map(&:user_name).and} said #{taunt.map(&:user_name).and} sucks at ping pong!")
+          logger.info "TAUNT: #{client.owner} - #{taunter.user_name}"
+        else
+          client.say(channel: data.channel, text: "You aren't registered to play, please _register_ first.", gif: 'register')
+          logger.info "TAUNT: #{client.owner} - #{taunter.user_name}, failed, not registered"
+
```
Comment (dblock) [inline slack-gamebot/commands/taunt.rb]: Kill the extra space pls.

File: slack-gamebot/commands/taunt.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
+module SlackGamebot
+  module Commands
+    class Taunt < SlackRubyBot::Commands::Base
+      def self.call(client, data, match)
+        taunter = ::User.find_create_or_update_by_slack_id!(client, data.user)
+        if taunter.registered?
+          arguments = match['expression'].split.reject(&:blank?) if match['expression']
+          arguments ||= []
```
Comment (dblock) [inline slack-gamebot/commands/taunt.rb]: `arguments = match['expression'] ? ... : []` is probably nicer, one-liner.

File: slack-gamebot/commands/taunt.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+          taunt = ::Taunt.create_from_teammates_and_opponents!(client.owner, data.channel, taunter, arguments)
+          client.say(channel: data.channel, text: "#{taunt.taunters.map(&:user_name).and} said #{taunt.map(&:user_name).and} sucks at ping pong!")
```
Comment (dblock) [inline slack-gamebot/commands/taunt.rb]: We support multiple games, it's not all ping pong, so should be `game.name`.

File: slack-gamebot/commands/taunt.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
+module SlackGamebot
+  module Commands
+    class Taunt < SlackRubyBot::Commands::Base
+      def self.call(client, data, match)
+        taunter = ::User.find_create_or_update_by_slack_id!(client, data.user)
+        if taunter.registered?
+          arguments = match['expression'].split.reject(&:blank?) if match['expression']
+          arguments ||= []
+          taunt = ::Taunt.create_from_teammates_and_opponents!(client.owner, data.channel, taunter, arguments)
```
Comment (dblock) [inline slack-gamebot/commands/taunt.rb]: What's this `Taunt` class? It doesn't seem to be included in the PR, but the spec passes because it's hiding the error.

File: spec/slack-gamebot/commands/taunt_spec.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
+require 'spec_helper'
+
+describe SlackGamebot::Commands::Taunt, vcr: { cassette_name: 'user_info' } do
+  let!(:team) { Fabricate(:team) }
+  let(:app) { SlackGamebot::Server.new(team: team) }
+  let(:client) { app.send(:client) }
+  let(:user) { Fabricate(:user, user_name: 'username') }
+  let(:opponent) { Fabricate(:user) }
+  it 'creates a singles taunt by user name' do
+    expect do
```
Comment (dblock) [inline spec/slack-gamebot/commands/taunt_spec.rb]: This `expect` is unnecessary, and actually hides the failure from this spec.

**Comment by dblock** [comment]
For CI failures run `rubocop -a ; rubocop --auto-gen-config` and make sure `rake` succeeds locally.

**Comment by marin-hyatt** [comment]
It's been a while since I committed this. As far as I know, all the checks have passed, and the test works. Is there anything else I need to do to make this better?

File: README.md (hunk 1 above)
Comment (dblock) [reply README.md]: This is incorrect, the code says "says".

**Comment by dblock** [comment]
You still need to update https://github.com/dblock/slack-gamebot/blob/master/CHANGELOG.md and https://github.com/dblock/slack-gamebot/blob/master/slack-gamebot/commands/help.rb, but I think that's it after that.

File: slack-gamebot/commands/taunt.rb [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+          client.say(channel: data.channel, text: 'Please only provide one user name to taunt.')
+        elsif arguments.empty?
+          client.say(channel: data.channel, text: 'Please provide a user name to taunt.')
+        else
+          victim = ::User.find_by_slack_mention!(client.owner, arguments.first)
+          if taunter.registered?
+            client.say(channel: data.channel, text: "#{taunter.user_name} says that #{victim.user_name} sucks at #{client.owner.game.name}!")
+            logger.info "TAUNT: #{client.owner} - #{taunter.user_name}"
+          else
+            client.say(channel: data.channel, text: "You aren't registered to play, please _register_ first.", gif: 'register')
```
Comment (dblock) [inline slack-gamebot/commands/taunt.rb]: I think this should move up. No need to lookup the person being taunted if the taunter isn't registered.

File: slack-gamebot/commands/taunt.rb [hunk 8]
```diff
@@ -0,0 +1,24 @@
+module SlackGamebot
+  module Commands
+    class Taunt < SlackRubyBot::Commands::Base
+      def self.call(client, data, match)
+        taunter = ::User.find_create_or_update_by_slack_id!(client, data.user)
+        arguments = match['expression'] ? match['expression'].split.reject(&:blank?) : []
+        if arguments.length > 1
```
Comment (dblock) [inline slack-gamebot/commands/taunt.rb]: Maybe this should support multiple people? Especially that we support doubles and such. It would mean looking up everyone, then joining them with an "and".

File: spec/slack-gamebot/commands/taunt_spec.rb [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+    expect(message: "#{SlackRubyBot.config.user} taunt <@#{victim.user_id}>", user: user.user_id).to respond_with_slack_message(
+      "#{user.user_name} says that #{victim.user_name} sucks at #{client.owner.game.name}!"
+    )
+  end
+  it 'taunts one person by user name' do
+    victim = Fabricate(:user, team: team)
+    expect(message: "#{SlackRubyBot.config.user} taunt #{victim.user_name}", user: user.user_id).to respond_with_slack_message(
+      "#{user.user_name} says that #{victim.user_name} sucks at #{client.owner.game.name}!"
+    )
+  end
```
Comment (dblock) [inline spec/slack-gamebot/commands/taunt_spec.rb]: You have code that handles multiple users, missing users, unregistered users. All those scenarios need tests.

**Comment by marin-hyatt** [comment]
I think I fixed all the problems you pointed out and also updated everything you asked me to. If you could look over this it would be great.

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,5 +1,6 @@
-### Changelog
+### Change log
 
+* [#153](https://github.com/dblock/slack-gamebot/pull/153): Added `taunt` command to enable users to taunt other users.
```
Comment (dblock) [inline CHANGELOG.md]: This needs your name in the end just like the rest of the lines.

**Comment by dblock** [comment]
Doesn't look like it. For one a bunch of my comments are still visible and this PR says there's a conflict in CHANGELOG.md. I think you need to push your changes up here (possibly force push with `-f`).

**Comment by marin-hyatt** [comment]
I made some commits so that taunt and taunt.spec pass all the tests, but the CI build still fails. It isn't failing because of my tests because it was failing before (without my command added), so I think that shouldn't be a problem for my commit.

**Comment by dblock** [comment]
Merged via https://github.com/dblock/slack-gamebot/commit/d46c7f1bf8bf9c9b76ec7eefa4762e9ceedb6398, thanks for hanging in here! Will get deployed soon.
